content/post/simulating-event-windows.md

```diff
@@ -0,0 +1,171 @@
+
+```
+2004-03-07 200 3
+2004-03-07 401 2
+```
+
+We will also adopt a small trick. Because we are simulating event-based windowing, we will introduce a "token" marker after each day, that our topology will have to interpret correctly. For example
+
+```
+64.242.88.10 - - [07/Mar/2004:16:24:16 -0800] "GET /twiki/bin/view/Main/ABlogPage HTTP/1.1" 200 4924
+64.242.88.10 - - [07/Mar/2004:16:29:16 -0800] "GET /twiki/bin/edit/Main/Header_checks?topicparent=Main.ConfigurationVariables HTTP/1.1" 401 12851
+64.242.88.10 - - [07/Mar/2004:16:30:29 -0800] "GET /twiki/bin/attach/Main/OfficeLocations HTTP/1.1" 401 12851
+END_OF_DAY
+64.242.88.10 - - [08/Mar/2004:06:11:48 -0800] "GET /twiki/bin/view/TWiki/AnotherPage HTTP/1.1" 200 3732
+64.242.88.10 - - [08/Mar/2004:06:12:50 -0800] "GET /twiki/bin/view/Main/YetAnotherPage HTTP/1.1" 200 40520
+```
+
+### Topology
+
+The simplest possible topology to do this would be composed of three elements, according to the following diagram
+
+![High Level Logfiles Diagram](/images/post/simulate-event-windowing-wallaroo/topology.png)
+
+A fully working version of the code from this post can be found in the [examples](http://github.com/) directory of our repository, together with instructions on how to build it and run it on some example data.
+
+We set up our Wallaroo topology as follows
+
+```
+def application_setup(args):
+    in_host, in_port = wallaroo.tcp_parse_input_addrs(args)[0]
+    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
+
+    ab = wallaroo.ApplicationBuilder("Apache Log file analysis")
+    ab.new_pipeline("convert",
+                    wallaroo.TCPSourceConfig(in_host, in_port, Decoder()))
+    ab.to(Count)
+    ab.to_sink(wallaroo.TCPSinkConfig(out_host, out_port, Encoder()))
+    return ab.build()
+```
+
+#### Decoder
+
+Our decoder needs to be able to correctly identify the `END_OF_DAY` message, and translate that into an appropriate message that will signal to the computations that they need to roll over to the next day. We will call this a `BoundaryMessage`, and all other messages will be `LogLine`s.
+
+```
+class Decoder(object):
+    def header_length(self):
+        return 4
+
+    def payload_length(self, bs):
+        return struct.unpack(">I", bs)[0]
+
+    def decode(self, bs):
+        if bs == 'END_OF_DAY':
+            return BoundaryMessage()
+        else:
+            return LogLine(bs)
+```
+
+#### Computation
+
+The computation itself needs to be able to do the following things:
+
+ - On the first `LogLine` for each window, determine which day we are counting for
+ - On all `LogLine`s, accumulate the counts
+ - On receipt of a `BoundaryMessage`, send all the output required as a
+   `SummaryMessage`
+
+```
+class Count(object):
+
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.current_batch = {}
+        self.current_day = None
+
+    def compute(self, data):
+        if type(data) is BoundaryMessage:
```

Comment:
Python nit: `isinstance(data, BoundaryMessage)` has been preferred over `type(foo) == Foo` in most python shops I worked for. I can't remember if it's in the official PEP8 stuff though.

Reply:
`isinstance` is generally preferred because it supports inheritance. In this particular program we don't have to worry about that because we aren't creating subclasses (and hey, composition over inheritance), but I think we should probably go with the more "pythonic" solution and use `isinstance`.

Having said that ...

This whole thing is going to need to be rewritten to use state, and when that's done I think one option would be to use a more OO solution like this:

```
class BoundaryMessage(object):
    # ...
    def process_with(self, computation, state):
        computation.process_boundary_message(self, state)
class LogLine(object):
    # ...
    def process_with(self, computation, state):
        computation.process_log_line(self, state)
class Count(object):
    def compute(self, data, state):
        return data.process_with(state)
    def process_boundary_message(data, state):
        r = state.current_batch
        state.reset()
        return r
    def process_current_line(data, state):
        return_code = determine_return_code(data)
        state.current_batch[return_code] = state.current_batch.get(return_code, 0) + 1
        if state.current_day is None:
            self.current_day = determine_day(data)  
```

Then again, Python programmers also hate classes, so I'm sure someone would say "HEY MAN, NICE JAVE LOL DON'T YOU READ STEVE YEGGEE?"

But I digress.

Reply:
Could also treat the log line as a string and do the test in the computation (since it's the computation that returns the output, this will work fine), in which case we don't end up with all the classes.

And I think the snark about python programmers` preferences is unfair.